### ISSUE TITLE
Are you still developing this?

### ISSUE THREAD

**Issue by ckaran** [magnet/socrates-rs#2]
I sincerely hope that you are still developing this, because it is something I could use right now...

**Comment by magnet** [comment]
Hey. The project is somewhat on hold, but I would like to get it to a usable state because it's so close and the hardest parts are done. Would be a waste to let it rot.

So, if you're interested and depending on your requirements on maturity levels, having some users might be an impetus to get to a first release.

**Comment by ckaran** [comment]
I have absolutely **no** requirements on maturity levels, beyond that it works to some degree.  If you need to change the API, I can adapt pretty easily; all of my code is for a simulator for my PhD, which is in constant flux, and which only has one user right now (me).

The main thing I need is clear documentation so I know how to use it, and how to adapt to changes whenever they happen.

**Comment by cedric-h** [comment]
Yeah, with some docs this would be amazing.

**Comment by magnet** [comment]
I have my hands full with other projects right now, but I hear you; I really need to fix some stuff before cutting a first release and adding doc. Help would be appreciated.

**Comment by cedric-h** [comment]
Let us know how we can help.
I see https://github.com/nagisa/rust_libloading/, it seems to do what is being done here at a much lower level, is that correct? Are there differences? It seems rust_libloading is still being actively developed. Is lib_loading something that socrates could depend on?

**Comment by magnet** [comment]
Yes Socrates is a dynamic service / module system built on top of libloading.

**Comment by cedric-h** [comment]
What is Socrates meant to do that LibLoading doesn't already provide?

**Comment by magnet** [comment]
Socrates provides a higher-level abstraction, think dependency injection and component reconfiguration. It's heavily inspired by OSGi, the dynamic module framework in Java. This lets you design your application in term of service dependencies, and if a required dependency is missing your component is in an unsatisfied state (and thus not called). It's ideal to write plugins that can reuse service provided by other plugins, or go all the way through and build the complete application as a set of interacting plugins, where the runner is basically an empty shell.

**Comment by cedric-h** [comment]
That sounds really awesome, how can we help?

**Comment by cedric-h** [comment]
If you could file some issues talking about specific things that people new to the crate could fix/add, that would be awesome.

**Comment by ckaran** [comment]
> It's ideal to write plugins that can reuse service provided by other plugins, or go all the way through and build the complete application as a set of interacting plugins, where the runner is basically an empty shell.

**THIS** is why I want to see this crate up and running 100%!  I want to have a set of easily replaced components in my simulator, so that you can choose your speed/accuracy trade-offs independent of what my framework provides.  So, echoing what @cedric-h suggested, list a set of issues that we as people new to the crate can start to look at.

**Comment by magnet** [comment]
There are two main blockers left:
- The `Svc<>` wrapper could be made safer; right now it is only safe if the container is dropped last, which could likely be guaranteed using lifetimes.
- The component system is not finished. This is the subsystem that tracks services going up and down dynamically and make the programming model simpler to reason about.

Nice to fix:
- The service layer could be decoupled from the module layer (e.g be useful even without libloading modules)
- The module entry point logic could be cleared
- Module unloading logic is not complete (but close, and it does not prevent coding)

Also this was my first serious Rust project and it could use a refresh of the code style and some light refactoring.

Unfortunately while I'm programming in Rust at work, we're not using the project because we don't really need hot-swappable components. I have little time for this right now, but I can provide guidance on gitter and feel free to open the issues for bullet points I mentioned. Hope this help :)

PS: the nice to have step would be adding support for WASM modules and consuming wasm or native code transparently ;)

**Comment by ckaran** [comment]
@magnet Hi, I saw your Gitter invite earlier, but I checked with our IT department, and they don't want me to use any IM clients.  So, GitHub issues are the only easy way to communicate with me.

**Comment by magnet** [comment]
> @magnet Hi, I saw your Gitter invite earlier, but I checked with our IT department, and they don't want me to use any IM clients. So, GitHub issues are the only easy way to communicate with me.

Hey @ckaran, no problem. In case you don't know gitter is a web app and doesn't require you install external clients or software, so it's similar to using GitHub issues to discuss, but in real-time.

**Comment by ckaran** [comment]
I understand, but the rules are the rules. So, GitHub issues it is!

**Comment by fayakun-it-consulting** [comment]
Ho guys I don't know if an attempt to revive this project with other folks might worth a try?

I actually contacted another project's owner who has an IoC crate in rust and asked if it make sense to add features from Socrates and complete the missing parts.

**Comment by fayakun-it-consulting** [comment]
I am well aware that there's this for C++

https://github.com/CppMicroServices/CppMicroServices

I used it many times and it does the job well and has also SCR implemented almost fully If I am not wrong!
